### ISSUE TITLE
BlockingConnection hangs for POLL_TIMEOUT after data is received

### ISSUE THREAD

**Issue by Ignalion** [pika/pika#698]
Hi, guys.

I've found the following issue.
Sometimes under high load of node where pika is worked (~80% CPU) and rather big message exchange (~1 message\sec) you could run into some sort of race condition that have unobvious (as for me) reasons.

A bit descriprion of environment:
pika runs in different thread (using threading.Thread) by Python compiled in some C++ application launched on Centos (I don't sure it matters but anyway).

I did some debugging and pika looks like to behave strangely.
General idea is: I simply run `channel.start_consuming()` and it should call callback provided by `channel.basic_consuming()` call.
But sometimes I see the following chain of calls:
* `PollPoller._poll.poll() call from PollPoller.poll() function, select_connection.py, line 590`
* `BaseConnection._handle_read() call from BaseConnection class, base_connection.py, line 364`
here we receive the data we need (i printed it from `connection.py, line 1343 _on_data_available`)
and again
* `PollPoller._poll.poll() call from PollPoller.poll() function, select_connection.py, line 590`
* **here we're waiting 5 seconds for timeout**
* and only now pika process data and call the callback function with it.

Could you please advise whether this is an expected behaviour for BlockingConnection (in fact, for me it looks a bit weird to say the least)?
As for me it's a bit weird to wait 5 seconds before processing the data we've already received
In fact, I think that switching to TwistedConnection as a workaround should resolve the issue because we get rid of blocking poll syscalls (could you please confirm?) and I'm considering this atm but anyway I'd like to have confirmation whether it's a bug or an expected behaviour

Thanks,
Constantine

**Comment by vitaly-krugl** [comment]
Hi @Ignalion, could you please attach a DEBUG-level pika log under these conditions? Also, just to be sure, are you by chance accessing the same pika connection (or any of its objects) from multiple threads?

**Comment by Ignalion** [comment]
Hi, @vitaly-krugl 
No, work with pika is being carried out only from one thread.
Regarding the logs. In fact there is not much logs to provide such as pika logging only its initialization.
I mean, you can't see any logs covering the issue time because pika does not do them. I did a bit of debugging so I can tell you what I'm seeing is.
Actually, I'm attaching startup pika log, I hope you found anything interesting there:

https://gist.github.com/Ignalion/dac49e941d246f6bacdc

**Comment by vitaly-krugl** [comment]
@Ignalion,  thanks for the logs. I confirm that this log doesn't show anything beyond initialization (setting up of the consumer). Clearly, processing of of messages should not depend on the 5-second timeout.

Next step is to instrument  with additional logging, reproduce the issue, and attach the new log (with timestamps, just as your previous log):

In `Connection._on_data_available`:
* log "_on_data_available" and the size of `data_in` on entry
* log `consumed_count` and `frame_value` immediately after `self._read_frame()` returns.

In `BlockingChannel._on_consumer_message_delivery`:
* On entry, log "_on_consumer_message_delivery", method and properties args, and length of body arg

In `BlockingChannel._dispatch_events`:
* On entry, log "_dispatch_events", channel number, and length of self._pending_events
* Inside the `if type(evt) is _ConsumerDeliveryEvt:` block, log `evt.method`, `evt.properties`, and length of `evt.body`

In `BlockingConnection._dispatch_channel_events`:
* Log "_dispatch_channel_events" and `self._channels_pending_dispatch` on entry
* Log "Nested dispatch or dispatch blocked higher in call stack" just before the `return` statement inside `if not dispatch_acquired:`

By the way, it would really help if you could submit a small, simple, fully-functional code sample that reproduces the issue.

**Comment by Ignalion** [comment]
@vitaly-krugl This is almost what I did for debugging it.

I got the logs you need especially for the message affected. There is some issue with loggers so I couldn't set a formatter for root logger unfortunately however I think the issue is pretty obvious here.

https://gist.github.com/Ignalion/4097c322a07bb65f8443

As you can see, all the data was received at one frame, the size of body matches.
There is 5 sec delay before _dispatch_channel_events method and I know exactly what it did during that time (I've seen it). It waits timeout for `events = self._poll.poll(self.get_next_deadline()` line.

**Comment by vitaly-krugl** [comment]
@Ignalion, did you instrument the log statement for "Nested dispatch or dispatch blocked higher in call stack"?

It would be best to attach a code sample for reproducing.

**Comment by Ignalion** [comment]
@vitaly-krugl yes, sure. It just didn't go under that if.

Regarding the code sample. Actually, I can provide you with the code however issue appears not every time so I can't tell exact circumstances for it to raise.

Here the class that handles requests by RabbitMQ:
https://gist.github.com/Ignalion/7c991cba7c18ecc8b938

**Comment by vitaly-krugl** [comment]
@Ignalion, I looked at your code snippets; don't see anything wrong with it. It must be BlockingConnection.

**Comment by vitaly-krugl** [comment]
@Ignalion, would you mind attaching more of that log. The 9 lines don't seem to be enough to figure out what state the connection and channel were in. The more, the better. Thx.

**Comment by Ignalion** [comment]
> From which thread/context will process_response be called? I suspect that it's going to be called from a different thread

No, the communication in this thread is doing only through queues. `process_response` will be called by retry function, here is it's code:
```
@inlineCallbacks
def retry(func, exc, interval=0.1, timeout=30):
    max_time = time() + timeout
    while time() <= max_time:
        try:
            result = func()
            returnValue(result)
        except exc:
            yield sleep(interval)
    else:
        raise TimeoutError
```

And if you interested, there is the code of sleep function as follows:
```
def sleep(t):
    awaken_later = Deferred()
    reactor.callLater(t, awaken_later.callback, None)
    return awaken_later
```

I guess what I'm trying to say is that the code of this snippet would be called only from current thread and it's absolutely thread safe.

**Comment by Ignalion** [comment]
@vitaly-krugl more logs for the logs god

https://gist.github.com/Ignalion/cd2de7ae4576eb613429

**Comment by vitaly-krugl** [comment]
@Ignalion: I created the following program that tries to emulate what your code is doing, but without the deferred business. So far, I've been running it on fedora and mac os x, but haven't been able to reproduce the issue. Would you mind taking a look and giving this a try in your environment. My goal is to have a "normal" BlockingConnection application (without twisted) that reproduces this issue.

```
import logging
import sys
import time

import pika


class IssueReproducer(object):

  def __init__(self):
    self.lastRxTime = None
    self.queue = None
    self.rxCount = 0


  def run(self):
    conn = pika.BlockingConnection()
    conn.add_on_connection_blocked_callback(self.onConnBlocked)

    ch = conn.channel()

    #ch.confirm_delivery()

    self.queue = ch.queue_declare(exclusive=True).method.queue

    ch.basic_publish(exchange="", routing_key=self.queue, body="1")
    ch.basic_publish(exchange="", routing_key=self.queue, body="2")

    self.lastRxTime = time.time()

    ch.basic_qos(prefetch_count=2)
    ch.basic_consume(self.onRx, queue=self.queue)

    ch.start_consuming()


  def onConnBlocked(self):
##    g_log.debug("ZZZ CONNECTION WAS BLOCKED")
    pass


  def onRx(self, ch, method, props, body):
    now = time.time()
    elapsed = now - self.lastRxTime
    self.lastRxTime = now

    self.rxCount += 1

##    g_log.debug("Rx # %s body=%s at %s elapsed=%s",
##      self.rxCount, body, now, elapsed)

    assert elapsed < 3, "rx message exceeded timeout; elapsed={}; body={}".format(
      elapsed, body)

    ch.basic_ack(delivery_tag=method.delivery_tag)

##    g_log.debug("Publishing body = %s", body)
    ch.basic_publish(exchange="", routing_key=self.queue, body=body)
##    g_log.debug("Done publishing body = %s", body)

##    time.sleep(0.2)


def main():
  global g_log

  LOG_FORMAT = ('%(levelname) -10s %(asctime)s %(name) -30s %(funcName) '
                '-35s %(lineno) -5d: %(message)s')
  logging.basicConfig(level=logging.DEBUG, format=LOG_FORMAT)

  g_log = logging.getLogger("TESTAPP")


  IssueReproducer().run()



if __name__ == "__main__":
  main()
```

**Comment by vitaly-krugl** [comment]
@Ignalion, also: try adding `%(thread)d` to your app's logging format, then add statement to log from your `process_response` closure (inside _on_request). After reproducing the problem, check the logs to see if the thread id of the `process_response` log ever differs from the thread id of the pika logs.

**Comment by Ignalion** [comment]
@vitaly-krugl Unfortunately I can't run your snippet in my environment because I can't block app where this code is running. It explains the different thread.
I tried though just run your code snippet and it just works w\o any exceptions. I ran it even during reproducing the issue with my code but this is expected.

In fact, I see this issue as race condition which is not easily reproducible. In your example you constantly sending two messages so issue most likely won't appear.
As far as I understand for the issue you should get only one message and then don't have any for timeout to starts.

In fact, the question here is a bit different. The question is why pika checking poll event AFTER it receives data. As for me this is the root cause.
That pika calling poll() when all system events where already fired.
Like, we have data we need and then we're trying to check whether there is any more data. If there is no system events, we have to wait 5 secs anyway before timeout is fired.

**Comment by vitaly-krugl** [comment]
@Ignalion, regarding
> UPD: I managed to reproduce the issue on my env even with running your script in parallel mode. It's very weird actually.

What do you mean by "running your script in **parallel mode**"?

**Comment by Ignalion** [comment]
@vitaly-krugl like I running the provided script and my script in different consoles:)

**Comment by vitaly-krugl** [comment]
@Ignalion, when you said that you were running it in "parallel mode", I thought you meant that you changed something in my script to make parts of it run in parallel. About how long did it take to reproduce the problem using my script?

In that case, you know the next thing I am going to ask you :) is the detailed DEBUG-level logs from reproducing the bug with my script with the following:
1. Uncomment all of the g_log statements in my script.
2. Instrumentation per https://github.com/pika/pika/issues/698#issuecomment-175149306

Also, along with the logs, please post the version of my script that you ran to reproduce the issue.

Many thanks.

**Comment by Ignalion** [comment]
@vitaly-krugl There is some misunderstanding.
Your script doesn't reproduce the issue. And in fact I have no idea how to adapt it for using in my env (actually I have but i'm thinking of trying to resolve the issue for starting get rid of multithreading).

In fact, I'd use Twisted Connection with async Twisted sockets but I can't do it because our app uses its own main loop so I can't just block it with reactor.run() call. So, I have to do some workarounds here.

**Comment by vitaly-krugl** [comment]
@Ignalion Got it! Since your app is in the Twisted environment, pika's Twisted adapters should be a better fit, and without any need for running a separate thread.

I don't have experience with Twisted and Pika's Twisted adapters, but it looks like you could submit a simple PR for pika's twisted_connection.py to make it work with an application-provided reactor.

The other asynchronous adapters in pika, such as SelectConnection and TornadoConnection, now accept an optional `ioloop` arg that overrides the default ioloop. So, you could add an additional optional `reactor` kwarg to the Twisted `TwistedConnection` and `TwistedProrocolConnection` to override the default one (similarly to how the other adapters process the optional `ioloop` arg). In that case, you don't need to call start/stop to get the adapter going if your app is already responsible for running that reactor. Together with `stop_ioloop_on_close=False`, this just might work!

You could probably easily try this out by patching `twisted_connection.reactor` with your app's custom reactor before instantiating `TwistedConnection` or `TwistedProrocolConnection`.

**Comment by vitaly-krugl** [comment]
@Ignalion, the script below (a specialized adaptation of your script) demonstrates that the pika connection is being accessed from another thread.

To reproduce:

1. Execute the script
2. When the log output to the terminal stops, interrupt execution via CTL-C

The last line on the terminal will show that a different thread was used, for example: `CRITICAL   2016-01-29 16:28:10,377 root                           process_response                     68  : process_response cmd_id=1 called from another thread=140735189172224; pika_thread=123145306509312`

I don't know Twisted enough (at all) to understand its interaction with Posix threads and why the deferred callback in my script is not getting executed automatically until I interrupt the program via CTL-C. It would still be cool to know why and how to get the callback to trigger automatically in my script, but not critical. The main point is that it shows that the mechanism can result in access to a pika connection from multiple threads.

```
DEBUG      2016-01-29 16:28:06,909 pika.callback                  process                              234 : Calling <bound method Channel._on_eventok of <pika.channel.Channel object at 0x100af0650>> for "1:Basic.ConsumeOk"
DEBUG      2016-01-29 16:28:06,909 pika.channel                   _on_eventok                          1000: Discarding frame <METHOD(['channel_number=1', 'frame_type=1', "method=<Basic.ConsumeOk(['consumer_tag=ctag1.2899148cf17649b9ab422f945e9c6fad'])>"])>
INFO       2016-01-29 16:28:06,909 root                           run                                  97  : RPC listener started
DEBUG      2016-01-29 16:28:06,909 root                           _on_request                          78  : RPC: got command 1
DEBUG      2016-01-29 16:28:06,909 root                           retry                                29  : Put response for command 1
DEBUG      2016-01-29 16:28:06,910 root                           _on_request                          85  : Leaving _on_request
^CDEBUG      2016-01-29 16:28:10,377 root                           retry                                26  : result is available now: 1
CRITICAL   2016-01-29 16:28:10,377 root                           process_response                     68  : process_response cmd_id=1 called from another thread=140735189172224; pika_thread=123145306509312
```


```python
import logging
from Queue import Queue, Empty
import os
import threading
from time import time

from twisted.internet import reactor
from twisted.internet.defer import inlineCallbacks, Deferred, returnValue, TimeoutError

import pika



def sleep(t):
  awaken_later = Deferred()
  reactor.callLater(t, awaken_later.callback, None)
  return awaken_later


@inlineCallbacks
def retry(func, exc, results_q, cmd_id, interval=0.1, timeout=30, ):
  max_time = time() + timeout
  while time() <= max_time:
    try:
      result = func()
      logging.debug("result is available now: %s", result)
      returnValue(result)
    except exc:
      logging.debug("Put response for command %s", cmd_id)
      results_q.put(cmd_id)
      yield sleep(interval)
  else:
    logging.critical("Timed out!!!")
    raise TimeoutError



class RPCServer(threading.Thread):


  def __init__(self, *args, **kwargs):
    super(RPCServer, self).__init__(*args, **kwargs)

    self.pika_thread_id = None

    self.command_id = 0

    self.connection = pika.BlockingConnection()
    self.channel = self.connection.channel()

    self.queue = self.channel.queue_declare(exclusive=True).method.queue

    self.results = Queue()
    self.daemon = True


  @property
  def _next_command_id(self):
    self.command_id += 1
    return str(self.command_id)


  def _on_request(self, ch, method, props, body):

    def process_response(response):
      if threading.current_thread().ident != self.pika_thread_id:
        logging.critical("process_response cmd_id=%s called from another thread=%s; pika_thread=%s",
                         response, threading.current_thread().ident, self.pika_thread_id)
        os._exit(1)

      next_id = self._next_command_id
      ch.basic_publish(exchange='',
                       routing_key=self.queue,
                       body=next_id)
      logging.debug("RPC: Got response=%s, published next command: %s", response, next_id)


    logging.debug('RPC: got command %s', body)

    ch.basic_ack(delivery_tag=method.delivery_tag)

    response_dfr = retry(self.results.get_nowait, exc=Empty, timeout=3600, results_q=self.results, cmd_id=body)
    response_dfr.addCallback(process_response)

    logging.debug("Leaving _on_request")


  def run(self):
    self.pika_thread_id = threading.current_thread().ident

    self.channel.basic_publish(exchange="", routing_key=self.queue, body=self._next_command_id)


    self.channel.basic_qos(prefetch_count=1)
    self.channel.basic_consume(self._on_request, queue=self.queue)

    logging.info('RPC listener started')
    self.channel.start_consuming()



def main():

  LOG_FORMAT = ('%(levelname) -10s %(asctime)s %(name) -30s %(funcName) '
                '-35s %(lineno) -5d: %(message)s')
  logging.basicConfig(level=logging.DEBUG, format=LOG_FORMAT)

  RPCServer().start()

  reactor.run()



if __name__ == "__main__":
  main()
```

**Comment by Ignalion** [comment]
Well, I get rid of threads here. I left blocking connection but simply get rid of threading and the issue seems to be resolved. I still don't understand how it works because I call and init pika and do all the communication with pika only from one different thread.
I think the socket and select operations in pika are not thread safe. And I think the possible explanation is that main thread also uses some poll() methods and because of this it's blocking pika's one.

**Comment by vitaly-krugl** [comment]
A pika Connection is not thread-safe. You can have multiple pika Connections, each accessed from its own Posix thread, just not shared between multiple threads.

It sounds like the issue is resolved, so closing it.
